stdlib/public/core/RangeSet.swift

```diff
@@ -0,0 +1,392 @@
+//===--- RangeSet.swift ---------------------------------------*- swift -*-===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2020 - 2023 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+/// A set of values of any comparable type, represented by ranges.
+///
+/// You can use a range set to efficiently represent a set of `Comparable`
+/// values that spans any number of discontiguous ranges. Range sets are
+/// commonly used to represent multiple subranges of a collection, by storing
+/// ranges of a collection's index type.
+///
+/// In this example, `negativeSubranges` is a range set representing the
+/// locations of all the negative values in `numbers`:
+///
+///     var numbers = [10, 12, -5, 14, -3, -9, 15]
+///     let negativeSubranges = numbers.subranges(where: { $0 < 0 })
+///     // numbers[negativeSubranges].count == 3
+///
+///     numbers.moveSubranges(negativeSubranges, to: 0)
+///     // numbers == [-5, -3, -9, 10, 12, 14, 15]
+@available(SwiftStdlib 5.11, *)
+public struct RangeSet<Bound: Comparable> {
+  internal var _ranges: Ranges
+
+  /// A collection of the ranges that make up the range set.
+  ///
+  /// The ranges that you access by using `ranges` never overlap, are never
+  /// empty, and are always in increasing order.
+  public var ranges: Ranges {
+    return _ranges
+  }
+
+  /// Creates an empty range set.
+  public init() {
+    _ranges = Ranges()
+  }
+  
+  /// Creates a range set containing the given range.
+  ///
+  /// - Parameter range: The range to use for the new range set.
+  public init(_ range: Range<Bound>) {
+    if !range.isEmpty {
+      self._ranges = Ranges(_range: range)
+    } else {
+      self._ranges = Ranges()
+    }
+  }
+  
+  /// Creates a range set containing the values in the given ranges.
+  ///
+  /// Any empty ranges in `ranges` are ignored, and non-empty ranges are merged
+  /// to eliminate any overlaps. As such, the `ranges` collection in the
+  /// resulting range set may not be equivalent to the sequence of ranges
+  /// passed to this initializer.
+  ///
+  /// - Parameter ranges: The ranges to use for the new range set.
+  public init(_ ranges: some Sequence<Range<Bound>>) {
+    self.init()
+    for range in ranges {
+      insert(contentsOf: range)
+    }
+  }
```

Comment:
It would probably make sense to mark this `@inlinable` -- it implements a trivial algorithm, and iteration over opaque sequence is not great. 

On a separate note, it would probably make sense to have a shortcut that forwards to `init(_ranges:)` when possible.

```suggestion
  @inlinable
  public init(_ ranges: some Sequence<Range<Bound>>) {
    if let ranges = _specialize(range, for: Ranges.self) {
      self.init(_ranges: ranges)
      return
    }
    self.init()
    for range in ranges {
      insert(contentsOf: range)
    }
  }
```

In fact, it would be even better if we just immediately converted `ranges` to an array and passed it to a new `Ranges` initializer that accepted nonsorted/overlapping contents (in-place sorting the array given and then collapsing its overlapping items).

```
  // Note: no longer a strong candidate for @inlinable
  public init(_ ranges: some Sequence<Range<Bound>>) {
    if let ranges = _specialize(range, for: Ranges.self) {
      self.init(_ranges: ranges)
      return
    }
    self.init(_ranges: Ranges(_unordered: Array(ranges))
  }
```

Reply:
Makes sense to me, I went with the latter approach and exposed an ABI entry point for initializing with an array of unordered ranges